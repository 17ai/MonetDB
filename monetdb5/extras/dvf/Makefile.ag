# The contents of this file are subject to the MonetDB Public License
# Version 1.1 (the "License"); you may not use this file except in
# compliance with the License. You may obtain a copy of the License at
# http://www.monetdb.org/Legal/MonetDBLicense
#
# Software distributed under the License is distributed on an "AS IS"
# basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
# License for the specific language governing rights and limitations
# under the License.
#
# The Original Code is the MonetDB Database System.
#
# The Initial Developer of the Original Code is CWI.
# Portions created by CWI are Copyright (C) 1997-July 2008 CWI.
# Copyright August 2008-2013 MonetDB B.V.
# All Rights Reserved.

INCLUDES = \
	../../optimizer \
	../../mal \
	../../../gdk \
	../../../common/stream \
	../../../common/options \
	$(MSEED_CFLAGS)

MTSAFE

lib_opt_dvf = {
	MODULE
	DIR = libdir/monetdb5
	SEP = _
	SOURCES = opt_dvf.c opt_dvf.h
	LIBS = ../../tools/libmonetdb5 \
		../../../gdk/libbat \
		../../../common/stream/libstream \
		$(MSEED_LIBS) \
		$(MALLOC_LIBS)
}

lib__dvf = {
	MODULE
	DIR = libdir/monetdb5
	SOURCES = dvf.c dvf.h
	LIBS = ../../tools/libmonetdb5 \
		../../../gdk/libbat \
		../../../common/stream/libstream \
		$(MSEED_LIBS) \
		$(MALLOC_LIBS)
}

headers_mal = {
	HEADERS = mal
	DIR = libdir/monetdb5
	SOURCES = dvf.mal opt_dvf.mal
}

headers_opt_dvf_autoload = {
	HEADERS = mal
	DIR = libdir/monetdb5/autoload
	SOURCES = 92_opt_dvf.mal
}

headers_dvf_autoload = {
	HEADERS = mal
	DIR = libdir/monetdb5/autoload
	SOURCES = 91_dvf.mal
}

EXTRA_DIST = 92_opt_dvf.mal dvf.mal 91_dvf.mal dvf.h

EXTRA_DIST_DIR = Tests
