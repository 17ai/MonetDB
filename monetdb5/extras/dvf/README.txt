==How to use mSEED data vault==

- libmseed library is needed.
	- Download it from http://www.iris.edu/pub/programs/libmseed-2.7.tar.gz
	- Install under LIBMSEED_LOCATION of your choice by referring to the file named 'INSTALL'. A simple 'make' on most Unix-like systems should build the library.
	- Create 'lib' and 'include' directories under LIBMSEED_LOCATION both pointing to the directory that contains 'libmseed.h'.

- MonetDB 'DVframework' branch is needed.
	- Refer to 'Building from Mercurial (HG) Sources' section of the wiki page http://www.monetdb.org/wiki/MonetDB:Building_from_sources
	- Check out the 'DVframework' branch by running "hg clone -u DVframework http://dev.monetdb.org/hg/MonetDB/"
	- Follow the steps on the wiki page and note that 'configure' should be run with this additional option: '--with-mseed=LIBMSEED_LOCATION'

- How to run an example
	- Go and Refer to .sql and .py files in the Tests/ directory under monetdb5/extras/dvf/.
	- You need a file containing a list of mSEED files with their absolute paths that you would like to have in the repository. Run "Mtest create_file_list" to get a small example of mSEED file repository.
	- Run "export LD_LIBRARY_PATH=$LIBMSEED_LOCATION/libmseed".
	- Run "mserver5" to start mserver with the default settings.
	- Run "mclient" in another terminal to get a client connection to the database server using 'demo' database.
	- In the mclient sql interface,
		- run 'mseed_schema.sql' script to create the normalized schema of mSEED.
		- run "CALL register_repo('TESTDIR/example_mseed_file_list.txt', 0);" where TESTDIR is the relative or absolute path of the aforementioned Tests/ directory. This will employ the registrar module to load (only) metadata of mSEED files.
	- Leave mclient and stop mserver5. Then run "mserver5 --readonly" to restart in readonly mode. The current incomplete version works only in readonly mode.
	- Run "mclient" again and in the mclient sql interface,
		- run 'initializer.sql' script to modify MonetDB optimizer pipeline in order to include DVframework optimizer in the pipeline. This is need to me done after every restart of mclient, unfortunately.
		- run example queries like "SELECT AVG(sample_value) FROM mseed.dataview WHERE network = 'FR';". You may also prefix the queries with 'EXPLAIN ' to see the query plan after the DVframework optimizer runs.
	- You may always refer to man pages of commands beginning with 'm' or 'M' for more information about them.
- E-mail to kargin@cwi.nl for questions, problems, etc.

