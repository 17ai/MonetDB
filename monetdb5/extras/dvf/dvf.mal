module dvf;

pattern plan_modifier{unsafe}(schema_name:str, file_locations:bat[:oid,:str]):void
address plan_modifier
comment "Modify the rest of the plan in the execution time according to the files to be mounted.";