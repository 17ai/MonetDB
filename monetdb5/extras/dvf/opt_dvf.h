/*
 * The contents of this file are subject to the MonetDB Public License
 * Version 1.1 (the "License"); you may not use this file except in
 * compliance with the License. You may obtain a copy of the License at
 * http://www.monetdb.org/Legal/MonetDBLicense
 *
 * Software distributed under the License is distributed on an "AS IS"
 * basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See the
 * License for the specific language governing rights and limitations
 * under the License.
 *
 * The Original Code is the MonetDB Database System.
 *
 * The Initial Developer of the Original Code is CWI.
 * Portions created by CWI are Copyright (C) 1997-July 2008 CWI.
 * Copyright August 2008-2015 MonetDB B.V.
 * All Rights Reserved.
 */

#ifndef _OPT_DVF_
#define _OPT_DVF_

#ifdef WIN32
#ifndef LIBOPT_DVF
#define opt_dvf_export extern __declspec(dllimport)
#else
#define opt_dvf_export extern __declspec(dllexport)
#endif
#else
#define opt_dvf_export extern
#endif

#include "opt_prelude.h"
#include "opt_support.h"

/*@:exportOptimizer(dvf)@*/
opt_dvf_export str OPTdvfIterative(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr p);
opt_dvf_export str OPTdvfSemiparallel(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr p);
opt_dvf_export str OPTdvfParallel(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr p);

//TODO: What are the following two lines?
/* #define DEBUG_OPT_DVF 61 
 # define OPTDEBUGdvf if (optDebug & ((lng)1 << DEBUG_OPT_DVF)) */
# define OPTDEBUGdvf if (optDebug)

#endif /* _OPT_DVF_ */

