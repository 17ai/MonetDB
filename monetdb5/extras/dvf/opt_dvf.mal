pattern optimizer.DVframeworkI():str
address OPTdvfIterative
comment "Inject a static MAL code to iteratively mount files, right after knowing about the files to be mounted.";

pattern optimizer.DVframeworkI(mod:str, fcn:str):str
address OPTdvfIterative
comment "Inject a static MAL code to iteratively mount files, right after knowing about the files to be mounted.";


pattern optimizer.DVframeworkS():str
address OPTdvfSemiparallel
comment "Inject an execution-time MAL plan modifier call right after knowing about the files to be mounted.";

pattern optimizer.DVframeworkS(mod:str, fcn:str):str
address OPTdvfSemiparallel
comment "Inject an execution-time MAL plan modifier call right after knowing about the files to be mounted.";


pattern optimizer.DVframeworkP():str
address OPTdvfParallel
comment "Inject an execution-time MAL plan modifier call right after knowing about the files to be mounted.";

pattern optimizer.DVframeworkP(mod:str, fcn:str):str
address OPTdvfParallel
comment "Inject an execution-time MAL plan modifier call right after knowing about the files to be mounted.";

