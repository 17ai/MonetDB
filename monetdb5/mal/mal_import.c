/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0.  If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 *
 * Copyright 1997 - July 2008 CWI, August 2008 - 2017 MonetDB B.V.
 */

/* Author(s) M.L. Kersten
 * Module import
 * The import statement simple switches the parser to a new input file, which
 * takes precedence. The context for which the file should be interpreted
 * is determined by the module name supplied.
 * Typically this involves a module, whose definitions are stored at
 * a known location.
 * The import context is located. If the module already exists,
 * we should silently skip parsing the file. This is handled at the parser level.
 * The files are extracted from a default location,
 * namely the DBHOME/modules directory.
 *
 * If the string starts with '/' or '~' the context is not changed.
 *
 * Every IMPORT statement denotes a possible dynamic load library.
 * Make sure it is loaded as well.
*/

#include "monetdb_config.h"
#include "mal_import.h"
#include "mal_builder.h"
#include "mal_interpreter.h"
#include "mal_linker.h"		/* for loadModuleLibrary() */
#include "mal_parser.h"
#include "mal_private.h"
#include "mal_exception.h"

/* #define _DEBUG_IMPORT_*/
void
slash_2_dir_sep(str fname)
{
	char *s;

	for (s = fname; *s; s++)
		if (*s == '/')
			*s = DIR_SEP;
}

static str
malResolveFile(str fname)
{
	char path[PATHLENGTH];
	str script;

	snprintf(path, PATHLENGTH, "%s", fname);
	slash_2_dir_sep(path);
	if ((script = MSP_locate_script(path)) == NULL) {
		/* this function is also called for scripts that are not located
		 * in the modpath, so if we can't find it, just default to
		 * whatever was given, as it can be in current dir, or an
		 * absolute location to somewhere */
		script = GDKstrdup(fname);
	}
	return script;
}

static stream *
malOpenSource(str file)
{
	stream *fd = NULL;

	if (file)
		fd = open_rastream(file);
	return fd;
}

/*
 * Beware that we have to isolate the execution of the source file
 * in its own environment. 
*/

#ifdef HAVE_EMBEDDED
extern char* mal_init_inline;
#endif

/*File and input processing
 * A recurring situation is to execute a stream of simple MAL instructions
 * stored on a file or comes from standard input. 
 * An Include operation simply pushes the input stream on the LIFO queue.
 */
str
evalFile(Client cntxt, str fname, int listing, int included)
{
	Client c;
	stream *fd;
	str p, filename, base;
	str files[MAXMULTISCRIPT];
	int cnt, i;
	str msg = MAL_SUCCEED;

	base = fname = malResolveFile(fname);
	if (fname == NULL) 
		throw(MAL,"mal.import", "#WARNING: could not open file: %s\n", fname);

	// load each individual file in the list
	// we should follow the sort order
	files[0] = filename = fname;
	cnt = 1;
	while ((p=strchr(filename, PATH_SEP)) != NULL) {
		*p = 0;
		//fprintf(stderr,"#loading %s\n", filename);
		filename = p+1;
		files[cnt++]= filename;
	}
	for(i=cnt-1; i>=0; i--){
		fd = malOpenSource(files[i]);
		if (fd == 0 || mnstr_errnr(fd) == MNSTR_OPEN_ERROR) {
			if(fd) mnstr_destroy(fd);
			msg = createException(MAL,"mal.import", "#WARNING: could not open file: %s\n", files[i]);
			GDKfree(base);
			return msg;
		} 

		if( included){
#ifdef _DEBUG_IMPORT_
			fprintf(stderr,"include file %s\n",files[i]);
#endif
			if( MCpushClientInput(cntxt, bstream_create(fd, 32 * BLOCK), listing, "")){
				GDKfree(base);
				throw(MAL,"mal.evalFile","Could not push the input stream");
			}
		} else {
#ifdef _DEBUG_IMPORT_
			fprintf(stderr,"load file %s using new client\n",files[i]);
#endif
			c = MCinitClient((oid)0,bstream_create(fd, 32 * BLOCK), cntxt->fdout);
			c->curmodule = c->usermodule = userModule();
			GDKfree(c->prompt);
			c->prompt= NULL;
			c->promptlength = 0;
			c->blkmode = 1; // collect all statements
			msg = defaultScenario(c);
			if( msg == MAL_SUCCEED){ 
				(void) MSinitClientPrg(c, "user", "main");  /* create new context */
				c->listing = listing;
				msg = runScenario(c);
			}
			c->fdout = NULL;	// to avoid accidental closing a default output channel
			MCcloseClient(c);
		}
	}
	GDKfree(base);
	return msg;
}
/*
 * The include operation simply pushes the file onto the preferred input queue.
 */
str
malInclude(Client c, str name, int listing)
{
	return evalFile(c, name,listing, 1);
}

/* patch a newline character if needed */
static str mal_cmdline(char *s, int *len)
{
	if (s[*len - 1] != '\n') {
		char *n = GDKmalloc(*len + 1 + 1);
		if (n == NULL)
			return s;
		strncpy(n, s, *len);
		n[*len] = '\n';
		n[*len + 1] = 0;
		(*len)++;
		return n;
	}
	return s;
}

/*
 * Compile the string, but don't execute it
 */
str
compileString(Client cntxt, str s)
{	int len = (int) strlen(s);
	buffer *b;
	str msg = MAL_SUCCEED;
	str qry, old=s ;
	Client c;

	if(0) s= mal_cmdline(s,&len);
	mal_unquote(qry = GDKstrdup(s));
	if( old != s)
		GDKfree(s);

    b = (buffer *) GDKzalloc(sizeof(buffer));
    if (b == NULL) {
        GDKfree(qry);
        return MAL_MALLOC_FAIL;
    }
    buffer_init(b, qry, len+2);

	c = MCinitClient((oid)0,0,0);
	c->fdin = bstream_create(buffer_rastream(b, "compileString"), b->len);
	strncpy(c->fdin->buf,s,len);
	c->curmodule = c->usermodule = userModule();
	GDKfree(c->prompt);
	c->prompt= NULL;
	c->promptlength = 0;

	msg = defaultScenario(c);
	if( msg == MAL_SUCCEED){ 
		(void) MSinitClientPrg(c, "user", "main");  /* create new context */
		c->blkmode = 1; // collect all statements
		while(msg == MAL_SUCCEED && c->fdin->eof == 0){
			msg = MALreader(c);
			if( msg == MAL_SUCCEED)
				msg = MALparser(c);
		}
		pushEndInstruction(c->curprg->def);
		chkProgram(c->usermodule, c->curprg->def);
	}
	c->fdout = 0;
	cntxt->curprg = c->curprg;
	c->curprg = 0;
	MCcloseClient(c);
	return MAL_SUCCEED;
}

str
callString(Client cntxt, str s)
{
	Client c;
	str msg = MAL_SUCCEED;

	c = MCinitClient((oid)0,cntxt->fdin,cntxt->fdout);
	msg = compileString(c,s);
	if( msg == MAL_SUCCEED)
		runMAL(c, c->curprg->def,0,0);
	c->fdin= 0;
	c->fdout = 0;
	MCcloseClient(c);
	return msg;
}
