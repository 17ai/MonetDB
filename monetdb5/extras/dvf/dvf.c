#include "monetdb_config.h"

#include <stdlib.h>
#include <stdio.h>
#include <unistd.h>

#include "../../modules/atoms/mtime.h"
#include "dvf.h"
#include "mal_interpreter.h"

int get_column_type(str schema_name, str table_name, int column_num)
{
	if(strcmp(schema_name, "mseed") != 0 || strcmp(table_name, "data") != 0)
		return -1;

	switch(column_num)
	{
		case 0:
			return TYPE_str;
		case 1:
			return TYPE_int;
		case 2:
			return TYPE_timestamp;
		case 3:
			return TYPE_int;
		default:
			return -1;
	}
}

int get_column_num(str schema_name, str table_name, str column_name)
{
	if(strcmp(schema_name, "mseed") != 0 || strcmp(table_name, "data") != 0)
		return -1;

	if(strcmp(column_name, "file_location") == 0)
		return 0;
	else if(strcmp(column_name, "seq_no") == 0)
		return 1;
	else if(strcmp(column_name, "sample_time") == 0)
		return 2;
	else if(strcmp(column_name, "sample_value") == 0)
		return 3;
	else
		return -1;
}

str plan_modifier(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr pci)
{
	str data_table_identifier = "data";
	str mountRef = putName("mount", 5);
	str miniseedRef = putName("miniseed", 8);
	str dvfRef = putName("dvf", 3);
	str planmodifierRef = putName("plan_modifier", 13);

	InstrPtr *old = NULL, *mounts = NULL, q = NULL, r = NULL, o = NULL;
	int i, j, k, limit, slimit, actions = 0;
	int num_fl = 0;
	BUN b1 = 0, b2 = 0;

	/* Declarations for copying of vars into stack and making a recursive runMALsequence call */
	str msg = MAL_SUCCEED;
	MalStkPtr stk_new;
	ValPtr lhs, rhs;
	int startpc = 1000000, old_vtop = 0;
	MalBlkPtr copy_old, copy_mb;
	bit is_stack_new = FALSE;

	BAT *BAT_fl = NULL; /* BAT for file_locations */

	bit after_first_data_bind = FALSE;

	str *schema_name = (str*) getArgReference(stk,pci,1); /* arg 1: schema_name */
	int bat_fl = *(int*) getArgReference(stk,pci,2); /* arg 2: bat of file_locations */
	int run_mergetable_opt = *(int*) getArgReference(stk,pci,3); /* arg 3: whether plan_modifier should integrate mergetable optimizer */

	BATiter fli;
	
	int run_dataflow_opt = 0;
	int run_recycle_opt = 1;

	VarRecord low, high;

	/* prepare to set low and high oids of return vars of mounts */
	high.value.vtype= low.value.vtype= TYPE_oid;
	high.value.val.oval = 0;
	low.value.val.oval = 0;

	/* check for logical error: mb must never be NULL */
	assert (mb != NULL);

	cntxt = cntxt; /* to escape 'unused' parameter error. */
	stk = stk; /* to escape 'unused' parameter error. */
	pci = pci; /* to escape 'unused' parameter error. */

	if ((BAT_fl = BATdescriptor(bat_fl)) == NULL)
		throw(MAL, "dvf.plan_modifier", RUNTIME_OBJECT_MISSING);

	/* check tail type */
	if (BAT_fl->ttype != TYPE_str)
	{
		throw(MAL, "dvf.plan_modifier",
		      "tail-type of input BAT must be TYPE_str");
	}

	BBPincref(bat_fl, TRUE);

	num_fl = BAT_fl->U->count;

	printf("files of interest: %d\n", num_fl);
	
	/* when number of files to be mounted is 0. */
	if(num_fl == 0)
		goto finish;

	/* prepare to keep the potential mount instructions */
	mounts = (InstrPtr*)GDKmalloc(num_fl*sizeof(InstrPtr));
	if(mounts == NULL)
		throw(MAL, "dvf.plan_modifier", MAL_MALLOC_FAIL);

	/* save the old stage of the MAL block */
	copy_mb = copyMalBlk(mb);
	copy_old = mb;
	mb = copy_mb;

	old = mb->stmt;
	limit= mb->stop;
	slimit = mb->ssize;


	/* initialize the statement list. Notice, the symbol table remains intact */
	if (newMalBlkStmt(mb, slimit) < 0)
		return 0;

	/* iterate over the instructions of the input MAL program, skip the dvf.plan_modifier itself. */
	for (i = 0; i < limit; i++)
	{
		InstrPtr p = old[i];

		/* check for
		 * dvf.plan_modifier(...);
		 */
		if(getModuleId(p) == dvfRef &&
			getFunctionId(p) == planmodifierRef)
		{
			startpc = i;
			pushInstruction(mb, copyInstruction(old[i]));
		}
		/* check for
		 * v7 := sql.bind(..., schema_name, data_table_name, ..., ...);
		 */
		else if(getModuleId(p) == sqlRef &&
			getFunctionId(p) == bindRef &&
			p->argc == 6 &&
			p->retc == 1 &&
			strcmp(getVarConstant(mb, getArg(p, 2)).val.sval, *schema_name) == 0 &&
			strstr(getVarConstant(mb, getArg(p, 3)).val.sval, data_table_identifier) != NULL)
		{
			int which_mount = 0;
			int which_column = 0;
			if(!after_first_data_bind)
			{
				int which_fl = 0;
				after_first_data_bind = TRUE;
				/* push mount instructions */
				/* create BAT iterator */
				fli = bat_iterator(BAT_fl);

				/* loop over the file_locations */

				BATloop(BAT_fl, b1, b2)
				{
					int a = 0, type;
					/* get tail value */
					str t = (str) BUNtail(fli, b1);

					/* create mount instruction */
					q = newInstruction(mb, ASSIGNsymbol);
					setModuleId(q, miniseedRef);
					setFunctionId(q, mountRef);
					for(; a < NUM_RET_MOUNT; a++)
					{
						type = get_column_type(*schema_name, getVarConstant(mb, getArg(p, 3)).val.sval, a);
						if(type < 0)
							throw(MAL, "dvf.get_column_num", "is not defined yet for schema: %s and table: %s and column: %s.", *schema_name, getVarConstant(mb, getArg(p, 3)).val.sval, getVarConstant(mb, getArg(p, 4)).val.sval);
						q = pushReturn(mb, q, newTmpVariable(mb, newBatType(TYPE_oid, type)));
						varSetProp(mb, getArg(q, a), PropertyIndex("hlb"), op_gte, (ptr) &low.value);
						varSetProp(mb, getArg(q, a), PropertyIndex("hub"), op_lt, (ptr) &high.value);
					}

					q = pushStr(mb, q, t);

					low.value.val.oval += 1;
					high.value.val.oval += 1;

					mounts[which_fl] = q;
					which_fl++;
					/* push the new instruction */
					pushInstruction(mb, q);
					actions++;
				}

				/* check for logical error */
				assert(which_fl == num_fl);

			}

			/* new mat.new for the column being binded */
			r = newInstruction(mb, ASSIGNsymbol);
			setModuleId(r, matRef);
			setFunctionId(r, newRef);
			r = pushReturn(mb, r, getArg(p, 0)); /* push the ret of sql.bind as ret of mat.new */
			which_column = get_column_num(*schema_name, getVarConstant(mb, getArg(p, 3)).val.sval,
						      getVarConstant(mb, getArg(p, 4)).val.sval);
			if(which_column < 0)
				throw(MAL, "dvf.get_column_num", "is not defined yet for schema: %s and table: %s and column: %s.",
				      *schema_name, getVarConstant(mb, getArg(p, 3)).val.sval, getVarConstant(mb, getArg(p, 4)).val.sval);
			for(; which_mount < num_fl; which_mount++)
			{
				r = pushArgument(mb, r, getArg(mounts[which_mount], which_column));
			}

			setVarInit(mb, getArg(r, 0));

			/* push the new instruction */
			pushInstruction(mb, r);
			actions++;

			/* arrange oids of return val of mat.new */
// 			s = newInstruction(mb, ASSIGNsymbol);
// 			setModuleId(s, algebraRef);
// 			setFunctionId(s, markHRef);
// 			s = pushReturn(mb, s, getArg(p, 0)); /* push the ret of sql.bind as ret of (mat.new + algebra.markH) */
// 			s = pushArgument(mb, s, getArg(r, 0));
// 			s = pushOid(mb, s, 0);

			/* push the new instruction */
// 			pushInstruction(mb, s);
// 			actions++;
/* 			s = s; */

			/* comment out in the old for reusing the old */
			copy_old->stmt[i]->token = REMsymbol;

		}
		else
		{
			/* push instruction */
			pushInstruction(mb, copyInstruction(old[i]));

			if (p->token == ENDsymbol) break;

			/* comment out in the old for reusing the old */
			if(i > startpc)
			{
/* 				old[i]->token = REMsymbol; */
				copy_old->stmt[i]->token = REMsymbol;
			}
		}
	}
	/* We would like to retain everything from the ENDsymbol
	 * up to the end of the plan, because after the ENDsymbol
	 * the remaining optimizer steps are stored.
	 */
	for(i++; i<limit; i++)
		if (old[i])
			pushInstruction(mb, copyInstruction(old[i]));


	/* save the vtop before calling any optimizer. */
	old_vtop = mb->vtop;

	/* call necessary optimizers
	 * *optimizer.inline();optimizer.remap();optimizer.evaluate();optimizer.costModel();optimizer.coercions();optimizer.emptySet();optimizer.aliases(); optimizer.mergetable();optimizer.deadcode();optimizer.commonTerms();optimizer.groups();optimizer.joinPath();optimizer.reorder();optimizer.deadcode();optimizer.reduce();optimizer.history();optimizer.multiplex();optimizer.accumulators();optimizer.garbageCollector();
	 */

	if(run_mergetable_opt)
	{
		o = newFcnCall(mb, "optimizer", "mergetable");
		typeChecker(cntxt->fdout, cntxt->nspace, mb, o, FALSE);
	}
	
	if(run_recycle_opt)
	{
		o = newFcnCall(mb, "optimizer", "recycle");
		typeChecker(cntxt->fdout, cntxt->nspace, mb, o, FALSE);
	}
	
	if(run_dataflow_opt)
	{
		o = newFcnCall(mb, "optimizer", "dataflow");
		typeChecker(cntxt->fdout, cntxt->nspace, mb, o, FALSE);
	}

	if(run_mergetable_opt || run_recycle_opt || run_dataflow_opt)
		optimizeMALBlock(cntxt, mb);

	/* New variables might have been created by the optimizers, so their values has to be copied into the stack. However, there might not be enough space in stack for them. We cannot reallocate the stack, but we may create our own enlarged stack, then run the rest of the plan with our own stack. */

	/* arrange the new stack without freeing the old one. */
	if (stk->stksize > mb->vsize)
		stk_new = stk;
	else
	{
		k = ((mb->vsize / STACKINCR) + 1) * STACKINCR;
		stk_new = newGlobalStack(k);
		memcpy(stk_new, stk, stackSize(stk->stksize));
		stk_new->stksize = k;
		is_stack_new = TRUE;
	}

	/* copy values into the new stack */
	for (j = 0; j < mb->vtop; j++) {
		lhs = &stk_new->stk[j];
		if (isVarConstant(mb, j) > 0) {
			if (!isVarDisabled(mb, j)) {
				rhs = &getVarConstant(mb, j);
				VALcopy(lhs, rhs);
			}
		}
		else
		{
			if (j > old_vtop)
			{
				lhs->vtype = getVarGDKType(mb, j);
				lhs->val.pval = 0;
				lhs->len = 0;
			}
			else if(j > stk->stksize)
			{
				rhs = getVarValue(mb, j);
				VALcopy(lhs, rhs);
			}
			else
			{
				rhs = &stk->stk[j];
				VALcopy(lhs, rhs);
			}
		}
	}
	stk_new->blk = mb;

	/* adjust variable lifetimes */
	malGarbageCollector(mb);

	/*chkProgram(cntxt->fdout, cntxt->nspace, mb);
	printFunction(cntxt->fdout,mb, 0, LIST_MAL_ALL); */

	/* relocate the startpc, the instruction to proceed with the execution. Because it might be changed by the optimizers. */
	for (i = 0; i < limit; i++)
	{
		InstrPtr ip = mb->stmt[i];
		
		/* check for
		 * dvf.plan_modifier(...);
		 */
		if(getModuleId(ip) == dvfRef &&
			getFunctionId(ip) == planmodifierRef)
		{
			startpc = i;
		}
	}
	
	msg = msg;
	/* run rest of the plan */
	msg = runMALsequence(cntxt, mb, startpc+1, mb->stop, stk_new, stk_new, mb->stmt[startpc]);

	if(msg != MAL_SUCCEED)
	{
		throw(MAL, "dvf.plan_modifier", "From the recursive call: %s", msg);
	}

/* 	chkProgram(cntxt->fdout, cntxt->nspace, copy_old); */
/* 	printFunction(cntxt->fdout, copy_old, 0, LIST_MAL_EXPLAIN); */

	/* any remaining MAL instruction records are removed */
	for(i = 0; i<slimit; i++)
		if (old[i])
			freeInstruction(old[i]);

	GDKfree(old);
	freeMalBlk(mb);
	if(is_stack_new)
		freeStack(stk_new);

finish:
	/* for statistics we print if/how many patches have been made */
	DEBUGoptimizers
	printf("#dvf.plan_modifier: %d actions\n", actions);

	return MAL_SUCCEED;
}

