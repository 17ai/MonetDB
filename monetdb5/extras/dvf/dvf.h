#ifndef _DVF_H
#define _DVF_H

#ifdef WIN32
#ifndef LIBDVF
#define dvf_export extern __declspec(dllimport)
#else
#define dvf_export extern __declspec(dllexport)
#endif
#else
#define dvf_export extern
#endif

#include "mal.h"
#include "mal_client.h"
#include "mal_interpreter.h"
#include "mal_function.h"
#include "opt_prelude.h"

dvf_export str plan_modifier(Client cntxt, MalBlkPtr mb, MalStkPtr stk, InstrPtr p);

//TODO: What is the following line?
#define _DVF_DEBUG_

#endif /* _DVF_H */