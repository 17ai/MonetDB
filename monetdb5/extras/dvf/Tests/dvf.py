import sys
import os
try:
        from MonetDBtesting import process
except ImportError:
        import process

def client(input):
        c = process.client('sql',
                stdin = process.PIPE,
                stdout = process.PIPE,
                stderr = process.PIPE)
        out, err = c.communicate(input)
        sys.stdout.write(out)
        sys.stderr.write(err)

f = open(os.path.join(os.environ['TSTSRCDIR'], 'initializer.sql'), 'r')

script1 = f.read() + '''\
SELECT AVG(sample_value) FROM mseed.dataview WHERE network = 'FR';
SELECT MAX(sample_value) FROM mseed.dataview WHERE seq_no < 1000;
'''

def main():
        s = process.server(args = ["--set", "gdk_readonly=yes"],
                stdin = process.PIPE,
                stdout = process.PIPE,
                stderr = process.PIPE)
        client(script1)
        out, err = s.communicate()
        sys.stdout.write(out)
        sys.stderr.write(err)

if __name__ == '__main__':
        main()
