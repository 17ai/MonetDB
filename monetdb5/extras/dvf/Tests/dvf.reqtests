mseed_schema
create_file_list
registrar
