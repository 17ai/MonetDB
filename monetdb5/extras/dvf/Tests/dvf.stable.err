stderr of test 'dvf` in directory 'extras/dvf` itself:


# 17:17:56 >  
# 17:17:56 >  "/usr/bin/python" "dvf.py" "dvf"
# 17:17:56 >  

# builtin opt 	gdk_dbname = demo
# builtin opt 	gdk_dbfarm = /export/scratch2/kargin/Monet/dft_miniseed/PREFIX/DFT/var/monetdb5/dbfarm
# builtin opt 	gdk_debug = 0
# builtin opt 	gdk_vmtrim = yes
# builtin opt 	monet_prompt = >
# builtin opt 	monet_daemon = no
# builtin opt 	mapi_port = 50000
# builtin opt 	mapi_open = false
# builtin opt 	mapi_autosense = false
# builtin opt 	sql_optimizer = default_pipe
# builtin opt 	sql_debug = 0
# cmdline opt 	gdk_nr_threads = 0
# cmdline opt 	gdk_dbfarm = /export/scratch2/kargin/Monet/dft_miniseed/PREFIX/DFT/var/MonetDB
# cmdline opt 	mapi_open = true
# cmdline opt 	mapi_port = 37875
# cmdline opt 	monet_prompt = 
# cmdline opt 	mal_listing = 2
# cmdline opt 	gdk_dbname = mTests_extras_dvf
# cmdline opt 	gdk_readonly = yes

# 17:17:57 >  
# 17:17:57 >  "Done."
# 17:17:57 >  

