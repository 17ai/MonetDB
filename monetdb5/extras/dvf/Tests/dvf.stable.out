stdout of test 'dvf` in directory 'extras/dvf` itself:


# 17:17:56 >  
# 17:17:56 >  "/usr/bin/python" "dvf.py" "dvf"
# 17:17:56 >  

#SET optimizer = 'sequential_pipe';
#declare opt_pipe_def string; set opt_pipe_def  = (select def from optimizers() where name = optimizer);
#declare opt_pipe_def string; set opt_pipe_def  = (select def from optimizers() where name = optimizer);
#set optimizer = substring(opt_pipe_def, 0, length(opt_pipe_def)-length('optimizer.joinPath();optimizer.reorder();optimizer.deadcode();optimizer.reduce();optimizer.history();optimizer.multiplex();optimizer.garbageCollector();')) || 'optimizer.reorder();optimizer.deadcode();optimizer.reduce();optimizer.history();optimizer.multiplex();optimizer.DVframework();optimizer.garbageCollector();';
#SELECT AVG(sample_value) FROM mseed.dataview WHERE network = 'FR';
% mseed. # table_name
% L1 # name
% double # type
% 24 # length
[ -8023.4	]
#SELECT MAX(sample_value) FROM mseed.dataview WHERE seq_no < 1000;
% mseed.dataview # table_name
% L1 # name
% int # type
% 6 # length
[ 123496	]
# MonetDB 5 server v11.14.0
# This is an unreleased version
# Serving database 'mTests_extras_dvf', using 8 threads
# Compiled for x86_64-unknown-linux-gnu/64bit with 64bit OIDs dynamically linked
# Found 15.629 GiB available main-memory.
# Copyright (c) 1993-July 2008 CWI.
# Copyright (c) August 2008-2012 MonetDB B.V., all rights reserved
# Visit http://www.monetdb.org/ for further information
# Listening for connection requests on mapi:monetdb://lodz.ins.cwi.nl:37875/
# MonetDB/GIS module loaded
# MonetDB/JAQL module loaded
# MonetDB/SQL module loaded
num_rows: 25
num_rows: 75
#BBPdecref: range error 931
#BBPdecref: range error 930
num_rows: 25
num_rows: 75
num_rows: 1872
num_rows: 3760
num_rows: 90406
#BBPdecref: range error 953
#BBPdecref: range error 946
#BBPdecref: range error 949

# 17:17:57 >  
# 17:17:57 >  "Done."
# 17:17:57 >  

